### PR TITLE
add simple cache for wiki download

### PR DESCRIPTION
**Pull request by jay-hodgson**

### CODE REVIEW

File: src/main/java/org/sagebionetworks/web/server/servlet/MarkdownCacheRequest.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
+package org.sagebionetworks.web.server.servlet;
+
```
Comment (john-hill) [inline src/main/java/org/sagebionetworks/web/server/servlet/MarkdownCacheRequest.java:2]: different key

File: src/main/java/org/sagebionetworks/web/server/servlet/SynapseClientImpl.java [hunk 2]
```diff
@@ -1645,43 +1676,42 @@ public String updateV2WikiPageWithV1(String ownerId, String ownerType,
...
-        } catch (JSONObjectAdapterException e) {
-            throw new UnknownErrorException(e.getMessage());
+        if (key.getWikiPageId() == null) {
+            //asking for the root.  find the root id first
+        	org.sagebionetworks.client.SynapseClient synapseClient = createSynapseClient();
+            String rootWikiPage = getV2RootWikiId(synapseClient, key.getOwnerObjectId(), ObjectType.valueOf(key.getOwnerObjectType()));
+            key.setWikiPageId(rootWikiPage);
         }
+        WikiPageKey properKey = new WikiPageKey(key.getOwnerObjectId(), ObjectType.valueOf(key.getOwnerObjectType()), key.getWikiPageId());
+        MarkdownCacheRequest request = new MarkdownCacheRequest(properKey, null);
```
Comment (jay-hodgson) [inline src/main/java/org/sagebionetworks/web/server/servlet/SynapseClientImpl.java]: get the most recent version before asking for the page from the cache.

File: src/main/java/org/sagebionetworks/web/server/servlet/SynapseClientImpl.java [hunk 3]
```diff
@@ -144,6 +149,30 @@
 			SynapseMarkdownProcessor.getInstance();
 		}
 	
+	private Cache<MarkdownCacheRequest, String> wiki2Markdown = CacheBuilder.newBuilder()
+			.maximumSize(500)
```
Comment (jay-hodgson) [inline src/main/java/org/sagebionetworks/web/server/servlet/SynapseClientImpl.java]: change to 35

File: src/test/java/org/sagebionetworks/web/unitserver/SynapseClientImplTest.java [hunk 4]
```diff
@@ -1377,6 +1385,29 @@ public void testSubmitCertificationQuizResponse() throws RestServiceException, S
...
+	public void testMarkdownCacheWithVersion() throws Exception {
+		Cache<MarkdownCacheRequest, String> mockCache = Mockito.mock(Cache.class);
+		synapseClient.setMarkdownCache(mockCache);
+		String pageJson = "test only";
+		when(mockCache.get(any(MarkdownCacheRequest.class))).thenReturn(pageJson);
+		Mockito.when(mockSynapse.getVersionOfV2WikiPage(any(org.sagebionetworks.repo.model.dao.WikiPageKey.class), anyLong())).thenReturn(v2Page);
+		String actualResult = synapseClient.getVersionOfV2WikiPageAsV1(new WikiPageKey(entity.getId(), ObjectType.ENTITY.toString(), "12"), 5L);
+		assertEquals(pageJson, actualResult);
+		verify(mockCache).get(any(MarkdownCacheRequest.class));
 	}
```
Comment (jay-hodgson) [inline src/test/java/org/sagebionetworks/web/unitserver/SynapseClientImplTest.java:1418]: do a cache hit test (verify times()==1)

**Pull request merged**